transports/quic/src/connection/stream.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    io::{self},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{AsyncRead, AsyncWrite};
+
+/// A single stream on a connection
+pub struct Stream {
+    /// A send part of the stream
+    send: quinn::SendStream,
+    /// A receive part of the stream
+    recv: quinn::RecvStream,
+    /// Whether the stream is closed or not
+    close_result: Option<Result<(), ()>>,
+}
+
+impl Stream {
+    pub(super) fn new(send: quinn::SendStream, recv: quinn::RecvStream) -> Self {
+        Self {
+            send,
+            recv,
+            close_result: None,
+        }
+    }
+}
+
+impl AsyncRead for Stream {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        if let Some(close_result) = self.close_result {
+            if close_result.is_err() {
+                return Poll::Ready(Ok(0));
+            }
+        }
+        let read_result = futures::ready!(Pin::new(&mut self.recv).poll_read(cx, buf));
+        if read_result.is_err() {
+            return Poll::Ready(Ok(0));
+        }
```

Comment:
Why do we need this? It seems wrong to not forward the underlying error.

Reply:
It does look wrong but dcutr+relay-server example does not work without this line :shrug:

Reply:
Can you expand on why it does not work @kpp? What logs are you seeing?

Reply:
Run `RUST_LOG=debug cargo r -p relay-server-example -- --secret-key-seed 1 --port 15353` and `RUST_LOG=debug cargo r -p dcutr -- --mode listen --secret-key-seed 2 --relay-address /ip4/127.0.0.1/udp/15353/quic-v1/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X`.

relay logs:

```
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /ipfs/ping/1.0.0
BehaviourEvent: Sent { peer_id: PeerId("12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3") }
[2023-07-04T17:51:44Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /ipfs/id/1.0.0
[2023-07-04T17:51:44Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /ipfs/ping/1.0.0
BehaviourEvent: Received { peer_id: PeerId("12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3"), info: Info { public_key: PublicKey { publickey: Ed25519(PublicKey(compressed): 6b79c57e6a95239282c4818e96112f3f3a401ba97a564c23852a3f1ea5fc) }, protocol_version: "/TODO/0.0.1", agent_version: "rust-libp2p/0.43.0", listen_addrs: ["/ip4/10.1.1.54/tcp/37225", "/ip4/172.18.0.1/tcp/37225", "/ip4/172.28.0.1/tcp/37225", "/ip4/192.168.2.5/tcp/37225", "/ip4/127.0.0.1/tcp/37225", "/ip4/127.0.0.1/udp/47440/quic-v1", "/ip4/172.31.0.1/tcp/37225", "/ip4/172.17.0.1/tcp/37225"], protocols: [StreamProtocol { inner: Right("/ipfs/id/1.0.0") }, StreamProtocol { inner: Right("/ipfs/ping/1.0.0") }, StreamProtocol { inner: Right("/libp2p/circuit/relay/0.2.0/stop") }, StreamProtocol { inner: Right("/ipfs/id/push/1.0.0") }], observed_addr: "/ip4/127.0.0.1/udp/15353/quic-v1/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X" } }
[2023-07-04T17:51:44Z DEBUG libp2p_ping::handler] latency to 12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3 is 1ms
BehaviourEvent: Event { peer: PeerId("12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3"), connection: ConnectionId(1), result: Ok(1.451498ms) }
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /libp2p/circuit/relay/0.2.0/hop
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /libp2p/circuit/relay/0.2.0/hop
BehaviourEvent: ReservationReqAccepted { src_peer_id: PeerId("12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3"), renewed: false }
[2023-07-04T17:51:44Z DEBUG libp2p_ping::handler] Inbound ping error: Custom { kind: NotConnected, error: ConnectionLost(ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" })) }
[2023-07-04T17:51:44Z DEBUG libp2p_swarm] Connection closed with error IO(Custom { kind: Other, error: Connection(ConnectionError(ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" }))) }): Connected { endpoint: Listener { local_addr: "/ip4/0.0.0.0/udp/15353/quic-v1", send_back_addr: "/ip4/127.0.0.1/udp/47440/quic-v1" }, peer_id: PeerId("12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3") }; Total (peer): 0.
```

dcutr logs:

```
[2023-07-04T17:51:44Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /ipfs/id/1.0.0
[2023-07-04T17:51:44Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /ipfs/ping/1.0.0
[2023-07-04T17:51:44Z INFO  dcutr] Relay told us our public address: "/ip4/127.0.0.1/udp/47440/quic-v1"
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /ipfs/id/1.0.0
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /ipfs/id/1.0.0
[2023-07-04T17:51:44Z DEBUG libp2p_ping::handler] latency to 12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X is 2ms
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: confirming protocol: /ipfs/ping/1.0.0
[2023-07-04T17:51:44Z DEBUG multistream_select::listener_select] Listener: sent confirmed protocol: /ipfs/ping/1.0.0
[2023-07-04T17:51:44Z INFO  dcutr] Told relay its public address.
[2023-07-04T17:51:44Z DEBUG multistream_select::dialer_select] Dialer: Proposed protocol: /libp2p/circuit/relay/0.2.0/hop
[2023-07-04T17:51:44Z DEBUG multistream_select::dialer_select] Dialer: Received confirmation for protocol: /libp2p/circuit/relay/0.2.0/hop
[2023-07-04T17:51:44Z DEBUG libp2p_swarm] Listener ListenerId(3); Closed by Ok(()).
[2023-07-04T17:51:44Z DEBUG libp2p_swarm] Connection closed with error Handler(Left(Left(Left(Left(Apply(Right(Codec(Error(Custom { kind: ConnectionReset, error: Stopped(0) }))))))))): Connected { endpoint: Dialer { address: "/ip4/127.0.0.1/udp/15353/quic-v1/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X", role_override: Dialer }, peer_id: PeerId("12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X") }; Total (peer): 0.
```

Reply:
> `[2023-07-04T17:51:44Z DEBUG libp2p_swarm] Listener ListenerId(3); Closed by Ok(()).`

My suspicion is that we are not keeping the connection alive. In other words my suspicion is that all `ConnectionHandler`s of the client return `KeepAlive::No` at some point and thus the connection is prematurely closed.

I just tested the following patch with https://github.com/libp2p/punchr/pull/76:

``` diff
modified   transports/quic/src/connection/stream.rs
@@ -57,11 +57,7 @@ impl AsyncRead for Stream {
                 return Poll::Ready(Ok(0));
             }
         }
-        let read_result = futures::ready!(Pin::new(&mut self.recv).poll_read(cx, buf));
-        if read_result.is_err() {
-            return Poll::Ready(Ok(0));
-        }
-        Poll::Ready(read_result)
+        Pin::new(&mut self.recv).poll_read(cx, buf)
     }
 }
```

With the above, the Rust punchr has the same success rate as it does without.

The Rust punchr [keeps the connection alive at all times](https://github.com/libp2p/punchr/blob/c3d2982d032f2514ccb2128285d04180a1a1088d/rust-client/src/main.rs#L223-L229).

We need to fix the keep alive issue, though that is orthogonal to the error handling here. @kpp mind cherry-picking the above patch or merging pull request https://github.com/kpp/rust-libp2p/pull/29?

Reply:
Friendly ping @kpp any objections merging https://github.com/kpp/rust-libp2p/pull/29?

Reply:
The above pull request is merged. Resolving this conversation.